shared/js/background/classes/agentspoofer.es6.js

```diff
@@ -0,0 +1,145 @@
+        let accum = 0
+        for (const agent of agentList) {
+            if (pick <= accum + agent.frequency) {
+                this.needsRotation = false
+                selectedAgent = agent.agentString
+                break
+            }
+            accum += agent.frequency
+        }
+        return selectedAgent || this.realAgent
+    }
+
+    /**
+     * Filter agents according to the filter list
+     * and criteria such as keeping major / minor versions
+     * relatively close.
+     */
+    filterAgents (agents) {
+        const osMajorVariance = 0
+        const osMinorVariance = 3
+        const browserMajorVariance = 3
+        const browserMinorVariance = 1000
+        // Filter for version constraints
+        agents = agents.filter(agent =>
+            agent.osMajor >= Number(this.parsedAgent.os.major) - Number(osMajorVariance) &&
+            agent.osMajor <= Number(this.parsedAgent.os.major) + Number(osMajorVariance) &&
+            agent.osMinor >= Number(this.parsedAgent.os.minor) - Number(osMinorVariance) &&
+            agent.osMinor <= Number(this.parsedAgent.os.minor) + Number(osMinorVariance) &&
+            agent.versionMajor >= Number(this.parsedAgent.major) - Number(browserMajorVariance) &&
+            agent.versionMajor <= Number(this.parsedAgent.major) + Number(browserMajorVariance) &&
+            agent.versionMinor >= Number(this.parsedAgent.minor) - Number(browserMinorVariance) &&
+            agent.versionMinor <= Number(this.parsedAgent.minor) + Number(browserMinorVariance))
+        // Filter out any excluded agents
+        if (agentStorage.excludedAgents.length > 0) {
+            agents = agents.filter(agent => !agentStorage.excludedAgents.every(excludePattern => excludePattern.test(agent.agentString)))
+        }
+        // Don't include our current agent (so it should always rotate)
+        agents = agents.filter(agent => agent.agentString !== this.spoofedAgent)
+        return agents
+    }
+
+    /**
+     * Return true if we should spoof UA for this request
+     */
+    shouldSpoof (request) {
+        let tab = tabManager.get({ tabId: request.tabId })
+        // Only change the user agent header if the current site is not whitelisted
+        // and the request is third party.
+        if (!!tab && tab.site.whitelisted) {
+            return false
+        }
+        if (this.isFirstParty(this.getRootURL(request), request.url)) {
+            return false
+        }
+        const domain = tldts.parse(request.url).domain
+        if (agentStorage.excludedDomains.length > 0 && agentStorage.excludedDomains.every(excluded => domain === excluded)) {
+            return false
+        }
+        return true
+    }
+
+    /**
+     * Tests whether the two URL's belong to the same
+     * first party set.
+     */
+    isFirstParty (url1, url2) {
+        const first = tldts.parse(url1).domain
+        const second = tldts.parse(url2).domain
+        return first === second ||
```

Comment:
isn't our definition of first party too broad here? Do we have to include entities? 🤔

Reply:
I don't know if we _have_ to, but it might increase breakage where entities are doing something. Maybe it doesn't matter - I'll run some tests without entities and see if I find any breakage.

Reply:
`findParent` will return the entity for a url if you do to compare entities https://github.com/duckduckgo/duckduckgo-privacy-extension/blob/develop/shared/js/background/utils.es6.js#L30